nothing_cli/theatrics.py

```diff
@@ -35,46 +37,128 @@ def marquis(title, description):
+    in the `context` and `presets` fields of a Procedure.
+    In other words, it stores all the values needed to call .format() with
+    to display a step in the source procedure.
+
+    Any keyname beginning with a __ is 'lazy'. The user is prompted for that before
+    the first time it is referenced, then it's stored.
+    The user is prompted for values of keys with regular during __init__.
+    Values from `presets` are stored immediately."""
+
+    def __init__(self, procedure: Procedure):
+
+        self.procedure = procedure
+        self.store: Dict[str, str] = {}
+        self.requisite_names = set(
+            chain(
+                (next(iter(p.keys())) for p in procedure.presets),
+                (c.var_name for c in procedure.context),
+            )
+        )
+
+        for preset in procedure.presets:
+            k, v = next(iter(preset.items()))
+            self.store[k] = v
+
+        eager_context_items = [
+            item for item in procedure.context if not item.var_name[:2] == "__"
+        ]
+
+        for item in eager_context_items:
+            self.store[item.var_name] = self.prompt_for_value(item)
+
+    def prompt_for_value(self, item: ContextItem) -> str:
+        """Use ContextItem.prompt to get a value from the user"""
+
+        value = ask(item.prompt)
+
+        return value
+
+    def get_interpolations(self, step: Step) -> Dict:
+        """Returns the dictionary of kwargs the step needs for .format()"""
+
+        key_names = self.get_format_names(step.prompt)
+        if not key_names.issubset(self.requisite_names):
+            raise KeyError("your .yml is fucked")
+
+        for key in key_names:
+            if key not in self.store:
+                context: ContextItem = next(
+                    c for c in self.procedure.context if c.var_name == key
+                )
+                self.store[key] = self.prompt_for_value(context)
+
+        return {k: v for k, v in self.store.items() if k in key_names}
+
+    def get_format_names(self, text: str) -> Set[str]:
+        """Return the names of any variables mentioned in the templates of `string`"""
+
+        formatter = Formatter()
+
+        return set(
+            template_name
+            for _, template_name, _, _ in formatter.parse(text)
+            if template_name is not None
+        )
+
+
 def interactive_walkthrough(procedure: Procedure) -> None:
     """Interactively walk through a Procedure"""
 
     marquis(procedure.title, procedure.description)
 
-    context_dict = {}
-
-    if procedure.context:
-        for item in procedure.context:
-            context_value = typer.prompt(item.prompt)
-            context_dict[item.var_name] = context_value
+    store = InterpolationStore(procedure)
 
     typer.echo()
     for step in procedure.steps:
-        run_step(step, context_dict)
+        step_header = typer.style(
+            f"{glot['step_prefix']} {step.number}:",
```

Comment:
note for future self (i.e. drop a todo): Why exactly are `Step`s aware of their own number?